### PR TITLE
Upgrade to 4.x Version

### PR DESCRIPTION
**Pull request by JacobAGTyler**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -4,10 +4,11 @@
     "type": "cakephp-plugin",
     "license": "MIT",
     "require": {
-        "php": ">5.6",
-        "cakephp/cakephp": "^3.7"
+        "php": ">=7.2",
+        "cakephp/cakephp": "4.0.*"
```
Comment (challgren) [inline composer.json]: Please use ^4.0 as the constraint

**Review by challgren — CHANGES_REQUESTED** [review]
Looks good so far, We may need to support the newer middleware authentication.

File: composer.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     },
     "require-dev": {
         "cakephp/cakephp-codesniffer": "^3.0",
-        "phpunit/phpunit": "^8.0",
+        "phpunit/phpunit": "^8.0.4",
```
Comment (challgren) [inline composer.json]: Suggestion is to use ^8.5

File: .travis.yml [hunk 3]
```diff
@@ -21,27 +21,27 @@ matrix:
...
       env: CODECOVERAGE=1 DEFAULT=0
 
 before_script:
   - if [[ $PREFER_LOWEST != 1 ]]; then composer install --prefer-source --no-interaction ; fi
-  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
   - if [[ $PREFER_LOWEST == 1 ]]; then composer require --dev dereuromark/composer-prefer-lowest; fi
+  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
 
-  - if [[ $CHECKS != 1 ]]; then composer require phpunit/phpunit:"^5.7.27|^6.0"; fi
+  - if [[ $CHECKS != 1 ]]; then composer require phpunit/phpunit:"^8.0"; fi
```
Comment (challgren) [inline .travis.yml]: Use PHPUnit ^8.5

File: .travis.yml [hunk 4]
```diff
@@ -21,27 +21,27 @@ matrix:
...
 
-    - php: 7.2
+    - php: 7.3
       env: CODECOVERAGE=1 DEFAULT=0
 
 before_script:
   - if [[ $PREFER_LOWEST != 1 ]]; then composer install --prefer-source --no-interaction ; fi
-  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
   - if [[ $PREFER_LOWEST == 1 ]]; then composer require --dev dereuromark/composer-prefer-lowest; fi
+  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
```
Comment (challgren) [inline .travis.yml:40]: Don't change the order.

File: composer.json [hunk 5]
```diff
@@ -4,11 +4,12 @@
...
     "license": "MIT",
     "require": {
-        "php": ">5.6",
-        "cakephp/cakephp": "^3.7"
+        "php": ">=7.2",
+        "cakephp/cakephp": "^4.0.1"
     },
     "require-dev": {
         "cakephp/cakephp-codesniffer": "^3.0",
+        "phpunit/phpunit": "^8.0.4",
```
Comment (challgren) [inline composer.json]: Don't include PHPUnit in the composer.json as its installed with the CI commands. Will add phpunit setup scripts for testing locally.

**Review by challgren — CHANGES_REQUESTED** [review]
See requested changes and travis build should pass.

File: .travis.yml (hunk 4 above)
Comment (JacobAGTyler) [reply .travis.yml:40]: There was an issue with the 'semver' requirement from 'dereuromark/composer-prefer-lowest', which is not encountered if requested with '--prefer-lowest --prefer-stable'.

File: composer.json (hunk 5 above)
Comment (JacobAGTyler) [reply composer.json]: Will remove.

**Comment by JacobAGTyler** [comment]
> Looks good so far, We may need to support the newer middleware authentication.

I figured I'd update to 4.x first and then the middleware auth can be done in a separate feature branch.

File: .travis.yml (hunk 3 above)
Comment (challgren) [reply .travis.yml]: Use PHPUnit ^8.5

File: .travis.yml (hunk 4 above)
Comment (challgren) [reply .travis.yml:40]: Yeah I'm conversing with @dereuromark and he said use ^8.5 he knows there's issues with phpunit and semver

File: .travis.yml [hunk 6]
```diff
@@ -38,11 +38,12 @@ matrix:
...
 before_script:
   - if [[ $PREFER_LOWEST != 1 ]]; then composer install --prefer-source --no-interaction ; fi
-  - if [[ $PREFER_LOWEST == 1 ]]; then composer require --dev dereuromark/composer-prefer-lowest; fi
-  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
 
   - if [[ $CHECKS != 1 ]]; then composer require phpunit/phpunit:"^8.0.4"; fi
 
+  - if [[ $PREFER_LOWEST == 1 ]]; then composer require --dev dereuromark/composer-prefer-lowest; fi
+  - if [[ $PREFER_LOWEST == 1 ]]; then composer update --prefer-dist --no-interaction --prefer-lowest --prefer-stable ; fi
+
```
Comment (challgren) [inline .travis.yml]: This block needs to be moved to back line 41

File: .travis.yml [hunk 7]
```diff
@@ -7,9 +7,9 @@ services:
   - postgresql
 
 php:
-  - 5.6
-  - 7.0
+  - 7.2
   - 7.3
```
Comment (dereuromark) [inline .travis.yml]: This is not needed. It is also already double covered by postgres and sqlite :)

File: phpunit.xml.dist [hunk 8]
```diff
@@ -3,7 +3,6 @@
     colors="true"
     processIsolation="false"
```
Comment (dereuromark) [inline phpunit.xml.dist]: All those false ones can go away, as they are redundant.

File: tests/TestCase/Controller/Component/ImpersonateComponentTest.php [hunk 9]
```diff
@@ -178,10 +188,13 @@ public function testLoginConfiged()
 
     /**
      * @return void
-     * @expectedException \Cake\Database\Exception
+     *
+     * @throws \Exception
      */
     public function testUnloadableUserModal()
     {
+        $this->expectException('\Cake\Database\Exception');
```
Comment (dereuromark) [inline tests/TestCase/Controller/Component/ImpersonateComponentTest.php]: A good practice is to move this right before the line that will throw the exception.

Comment (JacobAGTyler) [reply tests/TestCase/Controller/Component/ImpersonateComponentTest.php]: Have moved all exception expectations to just before the login() method that throws them.

File: phpunit.xml.dist (hunk 8 above)
Comment (JacobAGTyler) [reply phpunit.xml.dist]: Have removed the two false options mentioned.

File: .travis.yml (hunk 7 above)
Comment (JacobAGTyler) [reply .travis.yml]: I think I've removed the correct lines. :)

Comment (dereuromark) [reply .travis.yml]: You removed the wrong line. 7.2 should stay, 7.3 was marked by me.

Comment (dereuromark) [reply .travis.yml]: Min/Max always.

File: .travis.yml (hunk 6 above)
Comment (JacobAGTyler) [reply .travis.yml]: I think this is now functional and gives the correct order to work for all combinations in the variables matrix.

File: .travis.yml (hunk 7 above)
Comment (JacobAGTyler) [reply .travis.yml]: Ah ok sorry. Will revert.

File: .travis.yml [hunk 10]
```diff
@@ -7,7 +7,7 @@ services:
   - postgresql
 
 php:
-  - 7.3
+  - 7.2
```
Comment (dereuromark) [inline .travis.yml:10]: Without the 7.4 in here too you lose the matrix. I really just pointed out to remove the middle line 7.3.

Comment (dereuromark) [reply .travis.yml:10]: Again: min+max

Comment (JacobAGTyler) [reply .travis.yml:10]: Sorry. Yes makes sense.

**Comment by JacobAGTyler** [comment]
Think all issues have been resolved. Probably good to go. :)

**Review by JacobAGTyler — COMMENTED** [review]
I think all change items are resolved.

File: .travis.yml (hunk 4 above)
Comment (JacobAGTyler) [reply .travis.yml:40]: Just saw this, reverted sequence and used ^8.5

**Pull request merged**